File: Backend/hrms_feb_node/hrms_feb_node/services/dashboard/dashboarddata.js
const express = require('express');
const { userdata, holidaydata, leavedata, teamleavedata, getteamdata, getEmployeeData } = require('../../repository/dashboard/dashboard_data');

async function userDashboarddata(user_id, year) {
  try {
    const userData = await userdata(user_id);
    const leaves = await leavedata(user_id, year);

    let holiday = [];
    let teamLeaves = [];
    let currentUserFullName = "";

    if (userData.length > 0) {
      const groupId = userData[0].holiday_group;
      const projectId = userData[0].project_id;

      // Holiday data
      holiday = await holidaydata(groupId);
      currentUserFullName = userData[0].userfullname;


      // change the holidaydate format using map
      holiday = holiday.map(h => {
        const dateObj = new Date(h.holidaydate);
        const formatedDate = dateObj.toLocaleDateString("en-GB", {
          day: '2-digit',
          month: "short",
        });
        return {  // Push changed format to holiday
          ...h.JSON,
          holidayname: h.holidayname,
          holidaydate: formatedDate.replace(/ /g, " ")
        }
      });


      // Get team members data
      let teamData = projectId ? await getteamdata(projectId, user_id) : [];
      const reportingTeam = await getEmployeeData(user_id);
      const teamUserIds = [
        ...new Set(
          [...teamData, ...reportingTeam].map(u => u.user_id)
        )
      ];

      // team member leave data
      teamLeaves = await teamleavedata(teamUserIds);

      teamLeaves = teamLeaves.map(l => {
        return {
          ...l.JSON,
          employeename: l['emp_sum_lve_req.userfullname'],
          from_date: l.from_date,
          to_date: l.to_date,
          leaveday: l.leaveday,
          leave_period: l.leave_period,
          prefix_id: l['emp_sum_lve_req.prefix_id']
        }
      });
    }

    return { success: true, leaves, holiday, teamLeaves, userFullName: currentUserFullName };
  } catch (error) {
    return { success: false, error: error.message };
  }

}

module.exports = { userDashboarddata };

File: frontend/hrms_feb_flutter/lib/features/dashboard/domain/entities/dashboard_data.dart
class Leaves {
  final double? empLeaveLimit;
  final double? usedLeaves;

  Leaves({this.empLeaveLimit, this.usedLeaves});

  factory Leaves.fromJson(Map<String, dynamic> json) {
    return Leaves(
      empLeaveLimit: (json['emp_leave_limit'] as num?)?.toDouble(),
      usedLeaves: (json['used_leaves'] as num?)?.toDouble(),
    );
  }
}

class Holiday {
  final String? holidayname;
  final String? holidaydate;

  Holiday({this.holidaydate, this.holidayname});

  factory Holiday.fromJson(Map<String, dynamic> json) {
    return Holiday(
      holidayname: json['holidayname'],
      holidaydate: json['holidaydate'],
    );
  }
}

class TeamLeaves {
  final String? employeeName;
  final String? fromDate;
  final String? toDate;
  final int? leaveday;
  final int? leavePeriod;
  final int? prefixId;

  TeamLeaves({this.employeeName, this.fromDate, this.toDate, this.leaveday, this.leavePeriod, this.prefixId});

  factory TeamLeaves.fromJson(Map<String, dynamic> json) {
    return TeamLeaves(
      employeeName: json['employeename'],
      fromDate: json['from_date'],
      toDate: json['to_date'],
      leaveday: json['leaveday'],
      leavePeriod: json['leave_period'],
      prefixId: json['prefix_id'],
    );
  }
}

class DashboardData {
  final Leaves? leaves;
  final List<Holiday>? holidays;
  final List<TeamLeaves>? teamLeaves;
  final String? userFullName;

  DashboardData({this.leaves, this.holidays, this.teamLeaves, this.userFullName});
}